File: backend/services/bd_service/domain/entities/track.py
from dataclasses import dataclass, field
from datetime import datetime, date
from typing import Optional, Dict, Any
from pydantic import BaseModel, Field
from typing import List, Literal

@dataclass
class Track:
    id: Optional[int]
    title: str
    author: str
    genre: str
    text: str
    emotion: str                    
    emotion_intensity: float
    emotion_components: List[Dict[str, Any]] 
    audio_features: Dict[str, Any]
    release_date: date
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    VALID_EMOTIONS = {'happiness', 'sadness', 'fear', 'anger', 'disgust', 'anticipation'}
    MAX_INTENSITY = 10.0
    MIN_INTENSITY = 0.0

    def validate(self) -> bool:
        if not self.title or len(self.title.strip()) < 1:
            return False
        if not self.author or len(self.author.strip()) < 1:
            return False
        if not self.genre or len(self.genre.strip()) < 1:
            return False
        if self.emotion not in self.VALID_EMOTIONS:
            return False
        if not (self.MIN_INTENSITY <= self.emotion_intensity <= self.MAX_INTENSITY):
            return False
        if not self.audio_features:
            return False
        required_features = {'tempo', 'energy', 'danceability', 'acousticness', 
                           'instrumentalness', 'valence', 'key', 'mode', 
                           'loudness', 'speechiness', 'duration'}
        if not all(feature in self.audio_features for feature in required_features):
            missing = [f for f in required_features if f not in self.audio_features]
            print(f"Missing required audio features: {missing}")
            return False
        return True

    def calculate_average_score(self) -> float:
        scores = []
        if 'energy' in self.audio_features:
            scores.append(self.audio_features['energy'])
        if 'danceability' in self.audio_features:
            scores.append(self.audio_features['danceability'])
        if 'valence' in self.audio_features:
            scores.append(self.audio_features['valence'])
        return sum(scores) / len(scores) if scores else 0.0

    def get_emotion_category(self) -> str:
        if self.emotion_intensity < 3.0:
            return "low"
        elif self.emotion_intensity < 7.0:
            return "medium"
        else:
            return "high"

File: backend/services/bd_service_go/internal/domain/track.go
package domain

import (
    "strings"
    "time"
)

var ValidEmotions = map[string]bool{
    "happiness": true, "sadness": true, "fear": true,
    "anger": true, "disgust": true, "anticipation": true,
}

type EmotionComponent struct {
    Emotion string  `json:"emotion"`
    Weight  float64 `json:"weight"`
}

type AudioFeatures struct {
    Tempo            float64 `json:"tempo"`
    Energy           float64 `json:"energy"`
    Danceability     float64 `json:"danceability"`
    Acousticness     float64 `json:"acousticness"`
    Instrumentalness float64 `json:"instrumentalness"`
    Valence          float64 `json:"valence"`
    Key              int     `json:"key"`
    Mode             int     `json:"mode"`
    Loudness         float64 `json:"loudness"`
    Speechiness      float64 `json:"speechiness"`
    Duration         float64 `json:"duration"`
}

type Track struct {
    ID                int                `json:"id"`
    Title             string             `json:"title"`
    Author            string             `json:"author"`
    Genre             string             `json:"genre"`
    Text              string             `json:"text"`
    Emotion           string             `json:"emotion"`
    EmotionIntensity  float64            `json:"emotion_intensity"`
    EmotionComponents []EmotionComponent `json:"emotion_components"`
    AudioFeatures     AudioFeatures      `json:"audio_features"`
    ReleaseDate       time.Time          `json:"release_date"`
    CreatedAt         *time.Time         `json:"created_at,omitempty"`
    UpdatedAt         *time.Time         `json:"updated_at,omitempty"`
}

func (t *Track) Validate() bool {
    if strings.TrimSpace(t.Title) == "" || len(t.Title) > 255 {
        return false
    }
    if strings.TrimSpace(t.Author) == "" || len(t.Author) > 255 {
        return false
    }
    if strings.TrimSpace(t.Genre) == "" || len(t.Genre) > 100 {
        return false
    }
    if !ValidEmotions[t.Emotion] {
        return false
    }
    if t.EmotionIntensity < 0 || t.EmotionIntensity > 10 {
        return false
    }
    return true
}

func (t *Track) CalculateAverageScore() float64 {
    sum := t.AudioFeatures.Energy + t.AudioFeatures.Danceability + t.AudioFeatures.Valence
    return sum / 3.0
}

func (t *Track) GetEmotionCategory() string {
    switch {
    case t.EmotionIntensity < 3.0:
        return "low"
    case t.EmotionIntensity < 7.0:
        return "medium"
    default:
        return "high"
    }
}

// Custom JSON marshalling для emotion_components и audio_features будет обрабатываться на уровне БД.